README.md

```diff
@@ -3,30 +3,233 @@
+
+#### <a name="setup-client-connection-timeout"></a> Connection Timeout
+
+Time in milliseconds to wait for a successful connection. The default value is: `1000`.
+
+```javascript
+new Kafka({
+  clientId: 'my-app',
+  brokers: ['kafka1:9092', 'kafka2:9092'],
+  connectionTimeout: 3000
+})
+```
+
+#### <a name="setup-client-default-retry"></a> Default Retry
+
+The `retry` option can be used to set the default configuration. The retry mechanism uses a randomization function that grows exponentially.
+
+Available options:
+
+* __maxRetryTime__ - Maximum wait time for a retry in milliseconds. Default: `30000`
+* __initialRetryTime__ - Initial value used to calculate the retry in milliseconds (This is still randomized following the randomization factor). Default: `300`
+* __factor__ - Randomization factor. Default: `0.2`
+* __multiplier__ - Exponential factor. Default: `2`
+* __retries__ - Max number of retries per call. Default: `5`
+
+```javascript
+new Kafka({
+  clientId: 'my-app',
+  brokers: ['kafka1:9092', 'kafka2:9092'],
+  retry: {
+    initialRetryTime: 100,
+    retries: 8
+  }
+})
+```
+
+#### <a name="setup-client-logger"></a> Logger
+
+KafkaJS has a built-in `STDOUT` logger which outputs JSON. It also accepts a custom log creator which allows you to integrate your favorite logger library.
+There are 5 log levels available: `NOTHING`, `ERROR`, `WARN`, `INFO`, and `DEBUG`. `INFO` is configured by default.
+
+__How to configure the log level?__
+
+```javascript
+const { Kafka, logLevel } = require('kafkajs')
+
+// Create the client with the broker list
+const kafka = new Kafka({
+  clientId: 'my-app',
+  brokers: ['kafka1:9092', 'kafka2:9092'],
+  logLevel: logLevel.ERROR
+})
+```
+
+The environment variable `KAFKAJS_LOG_LEVEL` can also be used and it has precedence over the configuration in code, example:
+
+```sh
+KAFKAJS_LOG_LEVEL=info node code.js
+```
+
+__How to create a log creator?___
+
+A log creator is a function which receives a log level and returns a log function. The log function receives two arguments: namespace and log.
+
+`namespace` identifies the component which is performing the log, for example, connection or consumer.
+
+`log` is an object with the following keys: `level`, `timestamp`, 'logger', and `message`
+
+```javascript
+{
+  level: 'INFO', // NOTHING, ERROR, WARN, INFO, or DEBUG
+  timestamp: '2017-12-29T13:39:54.575Z',
+  logger: 'kafkajs',
+  message: 'Started',
+  // ... any other extra key provided to the log function
+}
+```
+
+The general structure looks like this:
+
+```javascript
+const MyLogCreator = logLevel => (namespace, log) => {
+  // Example:
+  // const { level, timestamp, logger, message, ...others } = log
+  // console.log(`${level} [${namespace}] ${message} ${JSON.stringify(others)}`)
+}
+```
+
+Example using [winston](https://github.com/winstonjs/winston):
+
+```javascript
+const winstom = require('winston')
+const toWinstomLogLevel = level => switch(level) {
+  case 'NOTHING':
```

Comment:
So the enum actually maps to numbers, in the end, I chose to expose the "label" to the creator because it allows the user to integrate with other libraries without knowing what is a 5 or a 4.